roles/satellite-clone/files/disassociate_capsules.rb

```diff
@@ -11,42 +12,39 @@ def prepare_hammer_cmd(command)
 end
 
 def run_hammer_cmd(command)
-  command = prepare_hammer_cmd(command)
-  `#{command}`
+  `#{prepare_hammer_cmd(command)}`
 end
 
-def get_info_from_hammer(command, column=1)
-  bash_parse = " | grep -v \"Warning:\" | tail -n+2 | awk -F, {'print $#{column}'}"
-  run_hammer_cmd(command + bash_parse)
+def get_ids_from_hammer(command, search = nil)
+  JSON.parse(run_hammer_cmd("--output=json " + command + (search ? " --search #{search}" : "")).map do |obj|
+    obj["Id"]
+  end
 end
 
 def capsule_lce_args(action, capsule_id, env)
   "--csv capsule content #{action}-lifecycle-environment --id #{capsule_id} --lifecycle-environment-id #{env}"
 end
 
-external_capsules = []
-external_capsule_ids = get_info_from_hammer("--csv capsule list --search 'feature = \"Pulp Node\"'")
-if external_capsule_ids.empty?
-  STDOUT.puts "There are no external capsules to disassociate."
-else
-  external_capsule_ids.split("\n").each do |id|
-    lifecycle_environment = get_info_from_hammer("--csv capsule content lifecycle-environments --id #{id}").split("\n")
-    name = get_info_from_hammer("--csv capsule info --id #{id}", 2).chomp
-    external_capsules << {:id => id, :name => name, :lifecycle_environments => lifecycle_environment}
-  end
+external_capsule_lifecycle_environments = get_ids_from_hammer("capsule list", search = 'feature = "Pulp Node"').map do |id|
+  { capsule_id: id, lifecycle_environments: get_ids_from_hammer("capsule content lifecycle-environments --id #{id}") }
+end
 
-  reverse_commands = []
-  external_capsules.each do |capsule|
-    capsule[:lifecycle_environments].each do |env|
-      run_hammer_cmd(capsule_lce_args("remove", capsule[:id], env))
-      reverse_commands << prepare_hammer_cmd(capsule_lce_args("add", capsule[:id], env))
-    end
+reverse_commands = external_capsule_lifecycle_environments.map do |association|
+  association[:lifecycle_environments].map do |env|
+    run_hammer_cmd(capsule_lcs_args("remove", association[:capsule_id], env))
+    prepare_hammer_cmd(capsule_lce_args("add", association[:capsule_id], env))
```

Comment:
I would keep this in the `else` part. While it's probably cheap, it's still something you don't need to do at all if `external_capsule_lifecycle_environments` is empty. The previous code wasn't poorly structured. There's certainly benefits to being able to decide very early on that you can abort execution.

There's a certain risk in blowing up the scope of a patch. While I certainly understand the desire to restructure things, it's important to find a good balance. The more you change, the more you can break.

Reply:
I got curious to know just how cheap or expensive is that loop when `external_capsule_lifeycycle_environments.empty?` or `external_capsule_lifecycle_environments.all? { |association| association[:lifecycle_environments].empty? }`

So I did a test:

```
irb(main):001:1* def timed_loop(somearray)
irb(main):002:1*   start = Time.now
irb(main):003:1*   
irb(main):004:2*   somearray.map do |elem|
irb(main):005:3*     elem[:some_key].map do |subelem|
irb(main):006:3*       "slept for #{sleep(60)} seconds"
irb(main):007:2*     end
irb(main):008:1*   end.flatten
irb(main):009:1*   
irb(main):010:1*   finish = Time.now
irb(main):011:1*   finish - start
irb(main):012:0> end
=> :timed_loop
irb(main):013:0> timed_loop([])
=> 2.259e-06
irb(main):014:0> timed_loop([{ some_key: [] }])
=> 5.11e-06
```

So it's on the order of a few microseconds. If I instead test something like 1000 "capsules" that each have no LCE assignments:

```
irb(main):036:1* myarray = 1000.times.map do |count|
irb(main):037:1*   { id: count, some_key: [] }
irb(main):038:0> end
=> [{:id=>0, :some_key=>[]}, {:id=>1, :some_key=>[]}, {:id=>2, :some_key=>[]}, {:id=>3, :some_key=>[]}, {:id=>4, :some_key=>[]}, {:id=>5, :some_key=>[]}, {:id=>6, :some_key=...
irb(main):039:0> timed_loop(myarray)
=> 0.000413076
```

Then it gets up to almost half a millisecond on my machine.

I realize the execution time isn't the point (we are on the same page about that) but I was just curious to know how it behaves :)

Reply:
But as we can see, since `"slept for #{sleep(60)} seconds"` never gets called in the above tests, in an implicit sense the interior of the loop is already within that if/else (speaking about the logic), the only difference being that we get `reverse_commands = []` instead of having it be undefined.

And I rather like that, because using `reverse_commands.empty?` seems like the best possible test for whether we need to print some output containing the reversal commands.

Reply:
This actually fixes a bug that existed previously, where we would try to print reversal commands and the full message about disassociation, when external capsules existed but their LCEs were already fully disassociated.

Fixing that bug otherwise would require changing the condition to `external_capsule_lifecycle_environments.all? { |association| association[:lifecycle_environments].empty? }` which is strictly worse because it still requires looping over `external_capsule_lifecycle_environments` once when they are all empty (same as my approach) but also requires looping over them twice (once for the conditional, once again to do the disassociations and populate `reverse_commands`) when they are populated, whereas my solution requires that only once in all cases.